### ISSUE TITLE
GridDetailRow Remounting Issue

### ISSUE THREAD

**Issue by PaitoAnderson** [telerik/kendo-react#249]
## I'm submitting a...
* Suggestion for improvement

## Current behavior
If you have an expanded row and you remove a row *above* it, the GridDetailRow component remounts losing any state.

## Expected behavior
It would be awesome if the GridDetailRow didn't remount losing the state you might be keeping in the GridDetailRow component.

## Minimal reproduction of the problem with instructions

Clicking Remove1 will cause a remount, clicking Remove2 won't.

```js
import React from 'react';
import ReactDOM from 'react-dom';
import { Grid, GridColumn as Column, GridToolbar, GridDetailRow } from '@progress/kendo-react-grid';

class DetailComponent extends GridDetailRow {
    componentDidMount() {
        console.log("componentDidMount")
    }
    render() {
        const dataItem = this.props.dataItem;
        return (
            <section>
                <p><strong>In Stock:</strong> {dataItem.UnitsInStock} units</p>
                <p><strong>On Order:</strong> {dataItem.UnitsOnOrder} units</p>
                <p><strong>Reorder Level:</strong> {dataItem.ReorderLevel} units</p>
                <p><strong>Discontinued:</strong> {dataItem.Discontinued}</p>
                <p><strong>Category:</strong> {dataItem.Category.CategoryName} - {dataItem.Category.Description}</p>
            </section>
        );
    }
}

class App extends React.Component {

    constructor(props) {
      this.state = {
        data: [{
            "ProductID" : 1,
            "ProductName" : "Chai",
            "SupplierID" : 1,
            "CategoryID" : 1,
            "QuantityPerUnit" : "10 boxes x 20 bags",
            "UnitPrice" : 18.0000,
            "UnitsInStock" : 39,
            "UnitsOnOrder" : 0,
            "ReorderLevel" : 10,
            "Discontinued" : false,
            "Category" : {
                "CategoryID" : 1,
                "CategoryName" : "Beverages",
                "Description" : "Soft drinks, coffees, teas, beers, and ales"
            }
        }, {
            "ProductID" : 2,
            "ProductName" : "Chang",
            "SupplierID" : 1,
            "CategoryID" : 1,
            "QuantityPerUnit" : "24 - 12 oz bottles",
            "UnitPrice" : 19.0000,
            "UnitsInStock" : 17,
            "UnitsOnOrder" : 40,
            "ReorderLevel" : 25,
            "Discontinued" : false,
            "Category" : {
                "CategoryID" : 1,
                "CategoryName" : "Beverages",
                "Description" : "Soft drinks, coffees, teas, beers, and ales"
            },
            "expanded": true
        }, {
            "ProductID" : 3,
            "ProductName" : "Aniseed Syrup",
            "SupplierID" : 1,
            "CategoryID" : 2,
            "QuantityPerUnit" : "12 - 550 ml bottles",
            "UnitPrice" : 10.0000,
            "UnitsInStock" : 13,
            "UnitsOnOrder" : 70,
            "ReorderLevel" : 25,
            "Discontinued" : false,
            "Category" : {
                "CategoryID" : 2,
                "CategoryName" : "Condiments",
                "Description" : "Sweet and savory sauces, relishes, spreads, and seasonings"
            }
        }]
      }

      this.removeTest1 = this.removeTest1.bind(this);
      this.removeTest2 = this.removeTest2.bind(this);
    }
    

    expandChange = (event) => {
        event.dataItem.expanded = !event.dataItem.expanded;
        this.forceUpdate();
    }

    render() {
        return (
            <Grid
            scrollable={'none'}
                data={this.state.data}
                detail={DetailComponent}
                style={{ height: '400px' }}
                expandField="expanded"
                onExpandChange={this.expandChange}
            >
            <GridToolbar>
              <button
                className="k-button k-primary"
                onClick={this.removeTest1}>
                Remove1
              </button>
              <button
                className="k-button k-primary"
                onClick={this.removeTest2}>
                Remove2
              </button>
            </GridToolbar>
                <Column field="ProductName" title="Product" width="300px" />
                <Column field="ProductID" title="ID" width="50px" />
                <Column field="UnitPrice" title="Unit Price" width="100px" />
                <Column field="QuantityPerUnit" title="Qty Per Unit" />
            </Grid>
        );
    }
    removeTest1() {
      this.setState((prevState) => ({ data: prevState.data.filter((x) => x.ProductID !== 1) })); 
    }
    removeTest2() {
      this.setState((prevState) => ({ data: prevState.data.filter((x) => x.ProductID !== 3) })); 
    }
}
ReactDOM.render(
    <App />,
    document.querySelector('my-app')
);
```

## What is the motivation or use case for changing the behavior?
We have a form in the GridDetailRow component that uses state, and we using SignalR to add/remove rows in realtime as other users collaborate on records which could result in a row being removed and users losing form data.

## Environment

Package versions:
```
├── @progress/kendo-data-query@1.5.1
├── @progress/kendo-drawing@1.5.12
├── @progress/kendo-react-buttons@2.11.0
├── @progress/kendo-react-dateinputs@2.11.0
├── @progress/kendo-react-dialogs@2.11.0
├── @progress/kendo-react-dropdowns@2.11.0
├── @progress/kendo-react-editor@2.11.0
├── @progress/kendo-react-grid@2.11.0
├── @progress/kendo-react-inputs@2.11.0
├── @progress/kendo-react-intl@2.11.0
├── @progress/kendo-react-layout@2.11.0
├── @progress/kendo-react-pdf@2.11.0
├── @progress/kendo-react-upload@2.11.0
├── @progress/kendo-theme-bootstrap@3.4.1
```

Browser:

- Chrome (desktop) version 74.0.3729.157

System:
- TypeScript version: 3.2.2
- Node version: v12.2.0 & v6.9.0
- Platform:  macOS

**Comment by PaitoAnderson** [comment]
My workaround for this was connecting components to redux to preserve state after remounting.

**Comment by Xizario** [comment]
Fixed in 4.2.0
